keps/262-ConfigurableFailurePolicy/README.md

```diff
@@ -0,0 +1,516 @@
+              values: [3]
+            reason: "ExitCode3"
+        template:
+          spec:
+            restartPolicy: Never
+            containers:
+            - name: main
+              image: python:3.10
+              command: ["..."]
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Proposed Failure Policy API
+
+```go
+
+// FailurePolicyAction defines the action the JobSet controller will take for
+// a given FailurePolicyRule.
+type FailurePolicyAction string
+
+const (
+  // Fail JobSet immediately, regardless of maxRestarts.
+  FailJobSet FailurePolicyAction = "FailJobSet"
+
+  // Don't count the failure against maxRestarts.
+  Ignore FailurePolicyAction = "Ignore"
+
+  // Restart all the child jobs of the ReplicatedJob the failed job is a part of.
+  RestartReplicatedJob FailurePolicyAction = "RestartReplicatedJob"
+  
+  // The failed child job is permanently failed and should not be
+  // restarted, active child jobs will continue to run to completion.
+  FailJob FailurePolicyAction = "FailJob"
+
+  // Restart the failed child job only, not the whole JobSet.
+  RestartJob FailurePolicyAction = "RestartJob"
+
+)
+
+// FailurePolicyRule defines a FailurePolicyAction to be executed if a child job
+// fails due to a reason listed in OnJobFailureReasons.
+type FailurePolicyRule struct {
+  // The action to take if the rule is matched.
+  // +kubebuilder:validation:Enum:=FailJobSet;Ignore;FailJob;RestartJob
+  Action FailurePolicyAction `json:"action"`
+  // The requirement on the job failure reasons. The requirement
+  // is satisfied if at least one reason matches the list.
+  // Each unique failure reason can only be associated with one FailurePolicyRule.
+  // An empty list applies the rule to any job failure reason.
+  // +kubebuilder:validation:UniqueItems:true
+  OnJobFailureReasons []string `json:"onJobFailureReasons"`
+  // TargetReplicatedJobs are the names of the replicated jobs the operator will apply to.
+  // A null or empty list will apply to all replicatedJobs.
+  // +optional
+  // +listType=atomic
+  // +kubebuilder:validation:UniqueItems
+  TargetReplicatedJobs []string `json:"targetReplicatedJobs,omitempty"`
+}
+
+type FailurePolicy struct {
+  // MaxRestarts defines the limit on the number of JobSet restarts.
+  // A restart is achieved by recreating all active child jobs.
```

Comment:
What if I have a rule for `RestartJob` or `RestartReplicatedJob`? Do those count towards this counter?

Do you already store the number of failures in the JobSet status? If not, that should be part of the KEP too.